### ISSUE TITLE
CPU and Memory Checks not working if windows ist installed with a non-english version

### ISSUE THREAD

**Issue by K0nne** [Icinga/icinga-powershell-plugins#132]
Hello,

it came to our attention that the checks `Invoke-IcingaCheckCPU` and `Invoke-IcingaCheckMemory` were not working on a specific system. Upon investigating the issue we found, that the system was installed with a german edition of windows server 2016. 

The problem seems to be, that even the performance counters are translated in german language:

**Processor**
```
icinga> Show-IcingaPerformanceCounterCategories -Filter '*Processor*'

icinga> Show-IcingaPerformanceCounterCategories -Filter '*Prozessor*'
Prozessor
Logischer Prozessor für Hyper-V-Hypervisor
Virtueller Prozessor des Hyper-V-Hypervisorstamms
Netzwerkschnittstellenkarten-Aktivität pro Prozessor
Netzwerkaktivitätszyklen pro Prozessor
Prozessorinformationen

icinga> Show-IcingaPerformanceCounters -Counter Prozessorinformationen
\Prozessorinformationen(*)\dpc-zeit (%)
\Prozessorinformationen(*)\privilegierte zeit (%)
\Prozessorinformationen(*)\dpcs in warteschlange/s
\Prozessorinformationen(*)\dpc-rate
\Prozessorinformationen(*)\% prozessorleistung
\Prozessorinformationen(*)\leerlaufzeit (%)
\Prozessorinformationen(*)\prozessorstatusflags
\Prozessorinformationen(*)\% c3-zeit
\Prozessorinformationen(*)\c3-übergänge/s
\Prozessorinformationen(*)\prozessorzeit (%)
\Prozessorinformationen(*)\interruptzeit (%)
\Prozessorinformationen(*)\% der maximalen frequenz
\Prozessorinformationen(*)\parkstatus
\Prozessorinformationen(*)\interrupts/s
\Prozessorinformationen(*)\benutzerzeit (%)
\Prozessorinformationen(*)\% c1-zeit
\Prozessorinformationen(*)\zeitinterrupts/s
\Prozessorinformationen(*)\prozessorauslastung
\Prozessorinformationen(*)\privilegierte auslastung
\Prozessorinformationen(*)\c2-übergänge/s
\Prozessorinformationen(*)\kennzeichen für das leistungslimit
\Prozessorinformationen(*)\durchschnittliche leerlaufzeit
\Prozessorinformationen(*)\prozessorfrequenz
\Prozessorinformationen(*)\unterbrechungen des leerlaufs/s
\Prozessorinformationen(*)\% leistungslimit
\Prozessorinformationen(*)\% prioritätszeit
\Prozessorinformationen(*)\% c2-zeit
\Prozessorinformationen(*)\c1-übergänge/s 
```

**Memory**
```
icinga> Show-IcingaPerformanceCounterCategories -Filter '*Arbeitsspeicher*'
Arbeitsspeicher
Arbeitsspeicher für NUMA-Knoten
Hyper-V-Dienst für die Integration des dynamischen Arbeitsspeichers

icinga> Show-IcingaPerformanceCounterCategories -Filter '*Memory*'
.NET Memory Cache 4.0

icinga> Show-IcingaPerformanceCounters -Counter Arbeitsspeicher
\Arbeitsspeicher\seitenlesevorgänge/s
\Arbeitsspeicher\zugesicherte bytes
\Arbeitsspeicher\freie und nullseitenlisten - bytes
\Arbeitsspeicher\nicht-auslagerungsseiten (bytes)
\Arbeitsspeicher\standbycache- bytes mit normaler priorität
\Arbeitsspeicher\cachebytes (max.)
```

In this case, the following message is thrown:

```
[UNKNOWN]: Icinga Invalid Configuration Error was thrown: PerfCounterCategoryMissing: Category "xxx" not found   

The specified Performance Counter category was not found on this system. This could either be a configuration error on your local Windows machine or a wrong usage of the plugin. Please check on different Windows machines if this issue persis. In case it only occurs on certain machines it is likely that the counter is simply not present and the plugin can not be processed.
```

Btw, there's a typo in the error message --> "persis"

**Comment by LordHepipud** [comment]
Thank you for the issue. How exactly was this system installed? My test machines are all installed by using the German localization and I never run into this issue so far.

It seems like the english counter part is not existing at all on the system, which makes me wonder on how other tools are querying the data required for runtime.

**Comment by K0nne** [comment]
The system was installed with a windows server 2016 and an additional german language pack.
The command seems to work just fine. Not sure why the checks are failing. They are executed by agent.
```
icinga> Show-IcingaPerformanceCounterInstances -Counter '\Processor(*)\% processor time';

Name                           Value
----                           -----
_Total                         \Processor(_Total)\% processor time
0                              \Processor(0)\% processor time
1                              \Processor(1)\% processor time


icinga>
```

Other people also stumbled over this in the past:
https://stackoverflow.com/questions/5581859/retrieve-performance-counter-value-in-a-language-independent-way

**Comment by LordHepipud** [comment]
I think here is a different issue happening. The problem is that the counters can be fetched just fine, because the way we implemented this always uses the englisch names.

How ever, we also added a test to check if a counter or category is available before doing anything else in some cases and the base code for this is using the same handling as `Show-IcingaPerformanceCounterCategories`

So in general this should mean the checks itself are running fine, but the test if a counter category is present or not fails because of this.

Right now I have no easy solution to fix this, as I'm checking for the category first (`Processor`) and break if this is not present, which oddly can happen on certain systems.

This will require additional investigation and will not be fixed with v1.4.0, but possibly with a later bugfix release.

**Comment by LordHepipud** [comment]
There is a new PR available on the Framework which will change on how the test for Performance Counter categories is done.

Could you please test this?

https://github.com/Icinga/icinga-powershell-framework/pull/248

**Comment by K0nne** [comment]
The PR works great! Now the checks are working.

**Comment by LordHepipud** [comment]
Resolved with https://github.com/Icinga/icinga-powershell-framework/pull/248

**Comment by log1-c** [comment]
hello :)

i'm not sure if I understand the fix correctly, but to me this does not make any sense, why it should solve the counter problem.
Because 

We currently have two identical Windows servers (Windows Terminal servers) installed with:

- Icinga2-v2.12.3-x86_64
- icinga-powershell-framework-1.4.1
- icinga-powershell-plugins-1.4.0
- icinga-service-v1.1.0
- Framework Code Cache enabled

Both are checked via the exe-files of the Agent normally, but for testing purposes we gave them both one cpu check via the powershell framework.

What was confusing is, that the check via powershell mostly worked on one of the systems, but not on the other.
![image](https://user-images.githubusercontent.com/24474580/120211684-7c540800-c231-11eb-8dcd-265f5f69973d.png)
Applying the fix did not solve the problem. The checks on one server would not work, the other did mostly.

Turns out that we had enabled the experimental API features on the mostly working machine before (for testing if this reduces the CPU load of the powershell framework checks).
Afteractivating the experimental features the checks started working.
![image](https://user-images.githubusercontent.com/24474580/120211846-a4dc0200-c231-11eb-9038-e6c7c2d942a3.png)
^ not working until API feature enabled
v working (mostly)
![image](https://user-images.githubusercontent.com/24474580/120211774-95f54f80-c231-11eb-8b5b-7f8b653aa218.png)

Maybe you can make sense of why the API feature makes the performance counter "detectable"?!

**Comment by LordHepipud** [comment]
Hello,

it shouldn't make a difference if system is checked by the API or the local handling. All that matters is, that the fix moved away by checking of the counter is present on the registry, but validates it by creating an object of the `Processor` category for example.

With the fix, both checks should work as you are able to call `Show-IcingaPerformanceCounterInstances`. What happens if you run `Test-IcingaPerformanceCounterCategory 'Processor'`.

On both systems, it should return `true`.

If not, what happens if you run

```powershell
New-Object System.Diagnostics.PerformanceCounterCategory('Processor');
```

on both systems.

**Comment by log1-c** [comment]
This is the return of both systems currently:
```
PS C:\Windows\system32> Test-IcingaPerformanceCounterCategory 'Processor'
False
PS C:\Windows\system32> New-Object System.Diagnostics.PerformanceCounterCategory('Processor')|fl


CategoryName : Processor
CategoryHelp : The Processor performance object consists of counters that measure aspects of processor activity. The
               processor is the part of the computer that performs arithmetic and logical computations, initiates
               operations on peripherals, and runs the threads of processes.  A computer can have multiple processors.
                The processor object represents each processor as an instance of the object.
CategoryType : MultiInstance
MachineName  : .
```

**Comment by LordHepipud** [comment]
Just to be sure - are you using the latest master of the Icinga PowerShell Framework and rebuild your Framework cache in case it is enabled with `Write-IcingaFrameworkCodeCache`.

This doesn't make any sense on why `Test-IcingaPerformanceCounterCategory` return false, while the object exists.

**Comment by log1-c** [comment]
sorry for the late answer.

The systems have v1.4.1 of the framework installed and I added the `Test-IcingaPerformanceCounterCategory` fix manually.
I'm not sure if if I explicitly ran the `Write-IcingaFrameworkCodeCache` afterwards, I assume I didn't^^

I just did the same procedure again, as the server got redeployed yesterday.
Manually edited the `Test-IcingaPerformanceCounterCategory` and ran the `Write-IcingaFrameworkCodeCache` afterwards.

```powershell
PS C:\Windows\system32> Write-IcingaFrameworkCodeCache
# in a new powershell
PS C:\Windows\system32> use-icinga
PS C:\Windows\system32> Test-IcingaPerformanceCounterCategory 'Processor'
True
```

I also have the second machine with the same setup but here I disabled the API features.
![image](https://user-images.githubusercontent.com/24474580/120478945-ad594780-c3ad-11eb-93b0-1ecc4245fab4.png)

so it looks like the fix is working :)
Though I will monitor the state changes now and check if they still occur.

**Comment by LordHepipud** [comment]
We just released Icinga for Windows 1.5.0 today, which natively includes this fix. It shouldn't occur again in the future.
